src/ppl/src/SimulatedAnnealing.cpp

```diff
@@ -163,66 +209,151 @@ int SimulatedAnnealing::getPinCost(int pin_idx)
   return netlist_->computeIONetHPWL(pin_idx, position);
 }
 
-void SimulatedAnnealing::perturbAssignment(int& prev_slot,
-                                           int& new_slot,
-                                           int& pin1,
-                                           int& pin2,
+int64 SimulatedAnnealing::getGroupCost(int group_idx)
+{
+  int64 cost = 0;
+  for (int pin_idx : pin_groups_[group_idx].pin_indices) {
+    int slot_idx = pin_assignment_[pin_idx];
+    const odb::Point& position = slots_[slot_idx].pos;
+    cost += netlist_->computeIONetHPWL(pin_idx, position);
+  }
+
+  return cost;
+}
+
+void SimulatedAnnealing::perturbAssignment(std::vector<int>& prev_slots,
+                                           std::vector<int>& new_slots,
+                                           std::vector<int>& pins,
                                            int& prev_cost)
 {
   const float move = distribution_(generator_);
 
   if (move < swap_pins_) {
-    prev_cost = swapPins(pin1, pin2);
+    prev_cost = swapPins(pins);
   } else {
-    prev_cost = movePinToFreeSlot(prev_slot, new_slot, pin1);
+    if (!pin_groups_.empty()) {
+      const float pin_or_group = distribution_(generator_);
+      if (pin_or_group <= move_groups_) {
+        prev_cost = moveGroupToFreeSlots(prev_slots, new_slots, pins);
+      } else {
+        prev_cost = movePinToFreeSlot(prev_slots, new_slots, pins);
+      }
+    } else {
+      prev_cost = movePinToFreeSlot(prev_slots, new_slots, pins);
+    }
   }
 }
 
-int SimulatedAnnealing::swapPins(int& pin1, int& pin2)
+int SimulatedAnnealing::swapPins(std::vector<int>& pins)
 {
-  pin1 = (int) (std::floor(distribution_(generator_) * num_pins_));
-  pin2 = (int) (std::floor(distribution_(generator_) * num_pins_));
+  int pin1 = (int) (std::floor(distribution_(generator_) * num_pins_));
+  int pin2 = (int) (std::floor(distribution_(generator_) * num_pins_));
   while (pin1 == pin2) {
     pin2 = (int) (std::floor(distribution_(generator_) * num_pins_));
   }
 
+  while (netlist_->getIoPin(pin1).isInGroup()) {
+    pin1 = (int) (std::floor(distribution_(generator_) * num_pins_));
+  }
+
+  while (netlist_->getIoPin(pin2).isInGroup()) {
+    pin2 = (int) (std::floor(distribution_(generator_) * num_pins_));
+  }
```

Comment:
Are you guaranteed to have to non-group pins?  

This could be slow if there are mostly pins in groups.  It might be better to separate them and then choose.  (Similarly in other places)

Is it not possible to swap a group with a block of whatever signals are in a set of slots?  If not it seems it will be hard to move groups as they are easily blocked by single bit pins.

Reply:
I agree with both suggestions. The second suggestion is more complex to implement because I need an efficient way to know what pin is placed in a specific slot. I can try to implement it in a future PR.